source/blog/_posts/2017-08-24-mutation-testing.md

```diff
@@ -0,0 +1,212 @@
+
+Mutation testing introduces a new quality score the so-called "Mutation Score Indicator". This score is the ratio of the number of Dead Mutants over all created Mutants. Usually this score is calculated like the code coverage in percent.
+
+In order to make this kind of testing automatically we can use [humbug](https://github.com/humbug/humbug) for that. Humbug has a wide range of mutators like the described mutations above. A good overview can be found [here](https://github.com/humbug/humbug#mutators). 
+
+So let us revert the new assertions and execute humbug for the first time. Humbug executes phpunit in the first place. After that it will create the mutants and executes the test suite again for every created mutant. To improve the execution time humbug only uses those test classes which cover the specific file and line on which the mutation was inserted.
+
+Humbug creates the following output: 
+```
+Humbug has completed the initial test run successfully.
+Tests: 2 Line Coverage: 100.00%
+
+Humbug is analysing source files...
+
+Mutation Testing is commencing on 1 files...
+(.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out)
+
+M.M.
+
+4 mutations were generated:
+       2 mutants were killed
+       0 mutants were not covered by tests
+       2 covered mutants were not detected
+       0 fatal errors were encountered
+       0 time outs were encountered
+
+Metrics:
+    Mutation Score Indicator (MSI): 50%
+    Mutation Code Coverage: 100%
+    Covered Code MSI: 50%
+```
+
+As we can see, humbug created 4 mutations, 2 mutants were killed and 2 mutants was not detected. So let us have a look at the generated mutations which are not detected:
+
+```php
+    public function isGreaterThan(int $x, int $y): bool
+    {
+        return $x >= $y;
+    }
+    
+    public function isSmallerThan(int $x, int $y): bool
+    {
+        return $x <= $y;
+    }    
+```
+
+Humbug detects automatically the same issues which we found above manually. If we add the new assertions which we already created above we should reach an Mutation Score Indicator of 100%. The created output stands below:
+```
+Humbug has completed the initial test run successfully.
+Tests: 2 Line Coverage: 100.00%
+
+Humbug is analysing source files...
+
+Mutation Testing is commencing on 1 files...
+(.: killed, M: escaped, S: uncovered, E: fatal error, T: timed out)
+
+....
+
+4 mutations were generated:
+       4 mutants were killed
+       0 mutants were not covered by tests
+       0 covered mutants were not detected
+       0 fatal errors were encountered
+       0 time outs were encountered
+
+Metrics:
+    Mutation Score Indicator (MSI): 100%
+    Mutation Code Coverage: 100%
+    Covered Code MSI: 100%
+```
+
+## Conclusion
```

Comment:
Die Floskel lautet imho "room for improvement", nicht "place for improvement" :)